src/renderer/atlas/BackendD2D.cpp

```diff
@@ -300,6 +308,138 @@ void BackendD2D::_drawText(RenderingPayload& p)
     }
 }
 
+f32 BackendD2D::_drawBuiltinGlyphs(const RenderingPayload& p, const ShapedRow* row, const FontMapping& m, f32 baselineY, f32 baselineX)
+{
+    const f32 cellTop = baselineY - p.s->font->baseline;
+    const f32 cellBottom = cellTop + p.s->font->cellSize.y;
+    const f32 cellWidth = p.s->font->cellSize.x;
+
+    _prepareBuiltinGlyphRenderTarget(p);
+
+    for (size_t i = m.glyphsFrom; i < m.glyphsTo; ++i)
+    {
+        u32 ch = row->glyphIndices[i];
+        if (til::is_leading_surrogate(ch))
+        {
+            i += 1;
+            ch = til::combine_surrogates(ch, row->glyphIndices[i]);
+        }
+
+        if (_builtinGlyphBatch)
+        {
+            if (const auto off = BuiltinGlyphs::GetBitmapCellIndex(ch); off >= 0)
+            {
+                const D2D1_RECT_F dst{ baselineX, cellTop, baselineX + cellWidth, cellBottom };
+                const auto src = _prepareBuiltinGlyph(p, ch, off);
+                const auto color = colorFromU32(row->colors[i]);
+                THROW_IF_FAILED(_builtinGlyphBatch->AddSprites(1, &dst, &src, &color, nullptr, sizeof(D2D1_RECT_F), sizeof(D2D1_RECT_U), sizeof(D2D1_COLOR_F), sizeof(D2D1_MATRIX_3X2_F)));
+            }
+        }
+
+        baselineX += row->glyphAdvances[i];
+    }
+
+    return baselineX;
+}
+
+void BackendD2D::_prepareBuiltinGlyphRenderTarget(const RenderingPayload& p)
+{
+    if (!_builtinGlyphBatch || _builtinGlyphsRenderTarget)

```

Comment:
```suggestion
    if (!_builtinGlyphBatch || !_builtinGlyphsRenderTarget)
```

(I'm only like half-understanding what's going on so correct me if I'm wrong) Wouldn't we want to exit early if `_buildinGlyphsRenderTarget` _is not_ defined? I see we overwrite it a few lines later. Curious.

Reply:
Ah, yeah that condition sucks. Let me split it apart and add comments!

Reply:
Should be much better now!